From dfba9d42a01a3a5cc879637bd716ce3c3f72fa0e Mon Sep 17 00:00:00 2001
From: Maksim Sisov <msisov@igalia.com>
Date: Tue, 29 Aug 2017 14:23:33 +0300
Subject: [PATCH] [ozone/x11] Add popup window type support.

This commit adds a popup window type to ozone windows.
Popup windows are drop down windows, for example.

Issue #417

fixup! Add popup window type support.

It added case handling on WindowTreeHostPlatform::GetWindowType().
---
 ui/platform_window/x11/x11_window_base.cc | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/ui/platform_window/x11/x11_window_base.cc b/ui/platform_window/x11/x11_window_base.cc
index e9a2cc3e5e7a8..51f5e8e17c380 100644
--- a/ui/platform_window/x11/x11_window_base.cc
+++ b/ui/platform_window/x11/x11_window_base.cc
@@ -79,15 +79,18 @@ void X11WindowBase::Create() {
   // type of the window.
   ui::PlatformWindowType ui_window_type = ui::PlatformWindowType::kWindow;
   // TODO(msisov, jkim): pass PlatformWindowInitProperties here.
-  if (ui_window_type != ui::PlatformWindowType::PLATFORM_WINDOW_TYPE_WINDOW) {
-    // Setting this to True, doesn't allow X server to set different
-    // properties, e.g. decorations.
-    // TODO(msisov): Investigate further.
-    // https://tronche.com/gui/x/xlib/window/attributes/override-redirect.html
-    swa.override_redirect = x11::True;
-    window_type = gfx::GetAtom("_NET_WM_WINDOW_TYPE_MENU");
-  } else {
-    window_type = gfx::GetAtom("_NET_WM_WINDOW_TYPE_NORMAL");
+   switch (ui_window_type) {
+    case ui::PlatformWindowType::kMenu:
+      window_type = gfx::GetAtom("_NET_WM_WINDOW_TYPE_MENU");
+      swa.override_redirect = x11::True;
+      break;
+    case ui::PlatformWindowType::kPopup:
+      swa.override_redirect = x11::True;
+      window_type = gfx::GetAtom("_NET_WM_WINDOW_TYPE_NOTIFICATION");
+      break;
+    default:
+      window_type = gfx::GetAtom("_NET_WM_WINDOW_TYPE_NORMAL");
+      break;
   }
 
   xwindow_ =
